src/app/Fake.IO.FileSystem/ChangeWatcher.fs

```diff
@@ -0,0 +1,112 @@
+[<AutoOpen>]
+/// This module contains helpers to react to file system events.
+module Fake.IO.FileSystem.ChangeWatcher
+
+open System.IO
+open Fake.Core
+open Fake.IO
+
+type FileStatus =
+    | Deleted
+    | Created
+    | Changed
+
+type FileChange =
+    { FullPath : string
+      Name : string
+      Status : FileStatus }
+
+type WatchChangesOption =
+    { IncludeSubdirectories: bool }
+
+let private handleWatcherEvents (status : FileStatus) (onChange : FileChange -> unit) (e : FileSystemEventArgs) =
+    onChange ({ FullPath = e.FullPath
+                Name = e.Name
+                Status = status })
+
+
+/// Watches the for changes in the matching files.
+/// Returns an IDisposable which allows to dispose all FileSystemWatchers.
+///
+/// ## Parameters
+///  - `onChange` - function to call when a change is detected.
+///  - `fileIncludes` - The glob pattern for files to watch for changes.
+///
+/// ## Sample
+///
+///     Target "Watch" (fun _ ->
+///         use watcher = !! "c:/projects/watchDir/*.txt" |> WatchChanges (fun changes ->
+///             // do something
+///         )
+///
+///         System.Console.ReadLine() |> ignore
+///
+///         watcher.Dispose() // if you need to cleanup the watches.
+///     )
+///
+let WatchChangesWithOptions options (onChange : FileChange seq -> unit) (fileIncludes : IGlobbingPattern) =
+    let dirsToWatch = fileIncludes |> GlobbingPattern.GetBaseDirectoryIncludes
+
+    //tracefn "dirs to watch: %A" dirsToWatch
+
+    // we collect changes in a mutable ref cell and wait for a few milliseconds to
+    // receive all notifications when the system sends them repetedly or sends multiple
+    // updates related to the same file; then we call 'onChange' with all cahnges
+    let unNotifiedChanges = ref List.empty<FileChange>
+    // when running 'onChange' we ignore all notifications to avoid infinite loops
+    let runningHandlers = ref false
+    let timerCallback = fun _ ->
+        lock unNotifiedChanges (fun () ->
+            if not (Seq.isEmpty !unNotifiedChanges) then
+                let changes =
+                    !unNotifiedChanges
+                    |> Seq.groupBy (fun c -> c.FullPath)
+                    |> Seq.map (fun (name, changes) ->
+                           changes
+                           |> Seq.sortBy (fun c -> c.Status)
+                           |> Seq.head)
+                unNotifiedChanges := []
+                try
+                    runningHandlers := true
+                    onChange changes
+                finally
+                    runningHandlers := false )
+    // lazy evaluation of timer in order to only start timer once requested
+    let timer = lazy( new System.Threading.Timer(timerCallback, System.Object(), 0, 50) )
```

Comment:
But we still need to `.Start()` the timer or not?

Reply:
I've added some more comments about it. Also changed so that it's more explicitly thread safe in the use of lazy.